packages/insured-bridge-relayer/src/ProfitabilityCalculator.ts

```diff
@@ -0,0 +1,236 @@
+import winston from "winston";
+import Web3 from "web3";
+const { toWei, toBN, fromWei } = Web3.utils;
+const toBNWei = (number: string | number) => toBN(toWei(number.toString()).toString());
+const fixedPoint = toBNWei(1);
+
+import { objectMap } from "@uma/common";
+import { Coingecko, across } from "@uma/sdk";
+import { getAddress } from "@uma/contracts-node";
+
+import { RelaySubmitType } from "./Relayer";
+
+import type { BN } from "@uma/common";
+
+export enum TokenType {
+  WETH,
+  ERC20,
+  UMA,
+}
+
+export class ProfitabilityCalculator {
+  public l1TokenInfo: { [token: string]: { tokenType: TokenType; tokenEthPrice: BN } } = {};
+
+  public relayerDiscount: BN;
+
+  private readonly coingecko;
+  /**
+   * @notice Constructs new Profitability Calculator Instance.
+   * @param {Object} logger Module used to send logs.
+   * @param {Array }l1Tokens list of L1 tokens the relayer is running on. Profitability module will store the prices of
+   * these tokens and use them when computing if a relay should be executed or not.
+   * @param {Number} l1ChainId L1 chain id that the relayer is running on.
+   * @param {Number} relayerDiscount Relayer discount applied to relays to offset gas cost.
+   */
+  constructor(
+    readonly logger: winston.Logger,
+    readonly l1Tokens: string[],
+    readonly l1ChainId: number,
+    readonly relayerDiscountNumber: number = 0
+  ) {
+    this.relayerDiscount = toBNWei(relayerDiscountNumber).div(toBN("100"));
+    if (this.relayerDiscount.lt(toBN(0)) || this.relayerDiscount.gt(toBNWei(1)))
+      throw new Error("Bad discount factor! 0<=relayerDiscount<=1e18");
+
+    this.coingecko = new Coingecko();
+  }
+
+  async update() {
+    this.logger.debug({ at: "ProfitabilityCalculator", message: "Updating prices", tokenList: this.l1Tokens });
+
+    // If the l1TokenInfo is empty then this is the first run. Figure out the token types of each l1Token.
+    if (Object.keys(this.l1TokenInfo).length == 0) {
+      const [umaAddress, wethAddress] = await Promise.all([
+        await getAddress("VotingToken", this.l1ChainId),
+        await getAddress("WETH9", this.l1ChainId),
+      ]);
+      for (const l1Token of this.l1Tokens) {
+        this.l1TokenInfo[l1Token] = { tokenType: TokenType.ERC20, tokenEthPrice: toBNWei("1") };
+        if (l1Token == umaAddress) this.l1TokenInfo[l1Token].tokenType = TokenType.UMA;
+        else if (l1Token == wethAddress) this.l1TokenInfo[l1Token].tokenType = TokenType.WETH;
+        else this.l1TokenInfo[l1Token].tokenType = TokenType.ERC20;
+      }
+    }
+
+    // Fetch the prices of each token, denominated in ETH. If coingecko does not have the price or is down then this
+    // call will fail and the tokenPrices array will have this error.
+    const tokenPrices = await Promise.allSettled(
+      this.l1Tokens.map((l1Token) => this.coingecko.getCurrentPriceByContract(l1Token, "eth"))
```

Comment:
1. Should we retry?
2. I think coingecko rate limits requests. Does this CG client make sure that we don't send requests too fast? Or do we need to manage that outside of the CG client?

Reply:
it is possible CG rate limits this. the client does not do any rate-limiting. I think it's fine for now (from my testing we can easily do 5 tokens at once). if we hit limits here we can slow down this call with a bluebird promise or add in retrys.

I added a comment to this effect.